File: src/network/packet/util/PacketSerializer.php
<?php

namespace pocketcloud\cloud\network\packet\util;

use JsonException;
use LogicException;
use pocketcloud\cloud\console\handler\ExceptionHandler;
use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\exception\PacketException;
use pocketcloud\cloud\network\packet\ClientboundPacket;
use pocketcloud\cloud\network\packet\CloudboundPacket;
use pocketcloud\cloud\network\packet\PacketPool;
use Throwable;

final class PacketSerializer {

    public static function encode(ClientboundPacket $packet, bool $encryptionEnabled, string $authenticationKey): ?string {
        try {
            $packet->encode($buffer = new PacketData());
            $buffer->write($authenticationKey);    $stringBuffer = json_encode($buffer, JSON_THROW_ON_ERROR);
            if ($encryptionEnabled) {
                $size = strlen($stringBuffer);
                if ($size > 10_000_000) CloudLogger::get()->warn("Huge packet before zlib: {} bytes, packet={}", $size, $packet->getName());
                $stringBuffer = zlib_encode($stringBuffer, ZLIB_ENCODING_DEFLATE, 3);
            }
            return $stringBuffer;
        } catch (Throwable $e) {
            throw new PacketException($e->getMessage(), $e->getCode(), $e);
        }
    }

    /**
     * @throws PacketException|JsonException
     */
    public static function decode(string $buffer, bool $encryptionEnabled, string $authenticationKey): ?CloudboundPacket {
        if ($buffer == "") throw new LogicException("Cannot decode an empty buffer");
        $data = json_decode($encryptionEnabled ? zlib_decode($buffer) : $buffer, true, flags: JSON_THROW_ON_ERROR);
        if (!is_array($data)) throw new PacketException("Received buffer is not an array");
        $packetName = $data[0] ?? null;
        if ($packetName === null) throw new PacketException("Received buffer does not contain a valid packet name");
        if (($packet = PacketPool::getInstance()->get($packetName)) !== null) {
            if (!$packet instanceof CloudboundPacket) throw new PacketException("Received packet is not a CloudboundPacket");
            $packet->decode($buffer = new PacketData($data));
            if ($buffer->isEmpty()) throw new PacketException("Received packet does not contain an authentication key");
            if (($givenKey = $buffer->readString()) === null) throw new PacketException("Received packet does not contain an authentication key");
            if ($givenKey !== $authenticationKey) throw new PacketException("Received packet does not contain a valid authentication key");
            return $packet;
        }

        return null;
    }
}

File: src/main/java/de/pocketcloud/cloud/network/packet/util/PacketSerializer.java
package de.pocketcloud.cloud.network.packet.util;

import com.google.gson.JsonSyntaxException;
import de.pocketcloud.cloud.network.packet.CloudPacket;
import de.pocketcloud.cloud.network.packet.PacketPool;
import de.pocketcloud.cloud.network.exception.PacketException;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.zip.DataFormatException;
import java.util.zip.Deflater;
import java.util.zip.Inflater;

public final class PacketSerializer {
    
    private PacketSerializer() {}

    public static byte[] encode(CloudPacket packet, boolean encryptionEnabled, String authenticationKey) throws PacketException {
        try {
            PacketData buffer = new PacketData();
            packet.encode(buffer);
            buffer.write(authenticationKey);
            
            String jsonBuffer = buffer.toJson();
            byte[] bytes = jsonBuffer.getBytes(StandardCharsets.UTF_8);
            
            if (encryptionEnabled) {
                bytes = compress(bytes);
            }
            
            return bytes;
        } catch (Exception e) {
            throw new PacketException(e.getMessage());
        }
    }

    public static CloudPacket decode(byte[] buffer, boolean encryptionEnabled, String authenticationKey) throws PacketException {
        try {
            if (buffer == null || buffer.length == 0) throw new PacketException("Cannot decode an empty buffer");
            byte[] decompressed = buffer;
            if (encryptionEnabled) {
                decompressed = decompress(buffer);
            }
            
            String json = new String(decompressed, StandardCharsets.UTF_8);
            PacketData data = PacketData.fromJson(json);
            
            if (data.isEmpty()) throw new PacketException("Received buffer is empty");
            
            String packetName = data.readString();
            if (packetName == null) throw new PacketException("Received buffer does not contain a valid packet name");
            
            var packet = PacketPool.getInstance().get(packetName);
            if (packet == null) return null;

            PacketData decodeData = PacketData.fromJson(json);
            packet.decode(decodeData);
            
            if (decodeData.isEmpty()) throw new PacketException("Received packet does not contain an authentication key");
            
            String givenKey = decodeData.readString();
            if (givenKey == null) throw new PacketException("Received packet does not contain an authentication key");
            
            if (!givenKey.equals(authenticationKey)) throw new PacketException("Received packet does not contain a valid authentication key");
            
            return packet;
        } catch (JsonSyntaxException e) {
            throw new PacketException("Failed to parse JSON: " + e.getMessage(), e);
        } catch (DataFormatException e) {
            throw new PacketException("Failed to decompress data: " + e.getMessage(), e);
        } catch (IOException e) {
            throw new PacketException("IO error during decoding: " + e.getMessage(), e);
        }
    }

    private static byte[] compress(byte[] data) throws IOException {
        Deflater deflater = new Deflater(3);
        try {
            deflater.setInput(data);
            deflater.finish();

            ByteArrayOutputStream outputStream = new ByteArrayOutputStream(data.length);
            byte[] buffer = new byte[1024];

            while (!deflater.finished()) {
                int count = deflater.deflate(buffer);
                outputStream.write(buffer, 0, count);
            }

            outputStream.close();
            return outputStream.toByteArray();
        } finally {
            deflater.end();
        }
    }

    private static byte[] decompress(byte[] data) throws IOException, DataFormatException {
        Inflater inflater = new Inflater();
        try {
            inflater.setInput(data);

            ByteArrayOutputStream outputStream = new ByteArrayOutputStream(data.length);
            byte[] buffer = new byte[1024];

            while (!inflater.finished()) {
                int count = inflater.inflate(buffer);
                outputStream.write(buffer, 0, count);
            }

            outputStream.close();
            return outputStream.toByteArray();
        } finally {
            inflater.end();
        }
    }
}